### PR TITLE
Stacker class

### PR DESCRIPTION
**Pull request by jaymccon**
## Overview

Adds a class to handle groups of `Stack` instances needed for `taskcat test` commands. For this purpose the `Stacker` class was ~~rewritten~~ created. It takes a `Config` instance as an input and can either `create_stacks` or import `from_existing` (to support decoupled/stateless use of taskcat in the future). Should additional deployment mechanisms (not cloudformation) be added in the future, they would need to implement only the public API's of `Template` and `Stacker` classes and should work as drop in replacements to the cloudformation implementation.

Operations that involve use of cloudformation API's are grouped into concurrency groups to ensure the highest level of concurrency while minimizing the risk of throttling. `Stack` properties and calls to `events`/`resources`/`children` use a cache (default ttl 1min) to make repeated calls fast and reduce api load.

this pr also sneaks in a change to bucket creation behavior, enforcing a particular region for each AWS partition, this was needed to unblock functional testing of nested stacks, and bring it inline with the QuickStart bucket strategy. I've opened #42 to track making this more flexible in the future.

## Testing/Steps taken to ensure quality

manual testing for now, would like to get some feedback and a `taskcat test` implementation done before spending effort on unit tests for what's likely to see a fair amount of short-term churn.

## Notes

There are some minor changes to other parts of taskcat included in this pr that were needed for functional testing to work.

Also squeezed in changes to catch the use of print statements, which can be handy during debugging, but we want to use logging module for all output

## Testing Instructions

setup a config:
```python
from pathlib import Path
from taskcat import Config

c = Config(
    project_config_path=Path('./tests/data/nested_fail/ci/taskcat.yml'
                             '').resolve(),
    project_root=Path('./tests/data/nested_fail/').resolve()
)
```

create the stacks:

```python
from taskcat._cfn.threaded import Stacker
from taskcat._s3_stage import stage_in_s3

stage_in_s3(c)
s = Stacker(c)
s.create_stacks()
```

Check the status:

```python
s.status()

Out[1]: 
{'IN_PROGRESS': {},
 'COMPLETE': {},
 'FAILED': {'arn:aws:cloudformation:us-west-2:...': 'The following resource(s) failed to create: [ChildStack]. ',
  'arn:aws:cloudformation:us-east-1:....': 'The following resource(s) failed to create: [ChildStack]. '}}
```

hmm, both stacks failed, let's look at what/why:

```python
for resources in s.resources(status='CREATE_FAILED').values():
    for resource in resources:
        print(f'{stack_id} resource {resource.logical_id} of type {resource.type} failed')
        
arn:aws:cloudformation:us-east-1:.... resource ChildStack of type AWS::CloudFormation::Stack failed
arn:aws:cloudformation:us-west-2:... resource ChildStack of type AWS::CloudFormation::Stack failed
```

Notice the filter being applied with `s.resources(status='CREATE_FAILED')` any property of an `Stack`, `Event` or `Resource` can be used as a filter. If multiple filters are passed they work as AND conditions.

Testing nested stacks is a bit involved, as the magic params `$[tasckcat_bucket_name]` have not been wired in yet. So performing the tests below involves setting a breakpoint just before stack launch to set the value of `QSS3Bucket` to the generated bucket name.

```
from pathlib import Path
from taskcat import Config
from taskcat._cfn.threaded import Stacker
from taskcat._s3_stage import stage_in_s3
c = Config(
    project_config_path=Path('./tests/data/nested-fail/ci/taskcat.yml'
                             '').resolve(),
    project_root=Path('./tests/data/nested-fail/').resolve()
)
stage_in_s3(c)
s = Stacker(c)
s.create_stacks()
```

after a bit you can see the child stack:
```
s.stacks[0].children()
Out[7]: [<Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-CRSYHC6ZUDVX at 0x10bbf7eb8>]
```

finding children of children and their descendants can be done easily:
```
s.stacks[0].descendants()
Out[8]: 
[<Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-CRSYHC6ZUDVX at 0x10bbf7eb8>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-Middle2Stack-L8E0QIBB0WI at 0x10c9fac18>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-C-ChildStack-15DCBOEQ195T0 at 0x10ca74fd0>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-M-ChildStack-19D1FG18JHOFK at 0x10ca74908>,
 <Stack object tCat-nested-fail-taskcat-json-1e2e315e06a94a9e8962b60347a228a7-ChildStack-M-ChildStack-10PR144QF0DDP at 0x10c3b9278>]
```

Once the stack fails:
```
s.stacks[0].status
Out[29]: 'CREATE_FAILED'
```
One can easily get the failure events without having to manually recurse through descendants:
```python
errors = s.stacks[0].error_events()
for error in errors:
    print(f"stack {error.stack_name} resource {error.logical_id} failed with error "
          f"{error.status_reason}")
```

outputs only the meaningful errors (not create cancelled or parent stack failures):
```
stack tCaT-nested-fail-taskcat-json-4600bbeff3d8479db842e1c79063dfd0-ChildStack-Y-ChildStack-K7TDJA8M86LP resource CustomResource failed with error Failed to create resource. See the details in CloudWatch Log Stream: 2019/08/19/[$LATEST]2ea497f9496740b29a6a8f9bc594b836
stack tCaT-nested-fail-taskcat-json-4600bbeff3d8479db842e1c79063dfd0-ChildStack-M-ChildStack-I6PBH6G7DGN4 resource CustomResource failed with error Failed to create resource. See the details in CloudWatch Log Stream: 2019/08/19/[$LATEST]cca9ad954e2e4f668685fb7ee692c65a
```

### CODE REVIEW

File: taskcat/_cfn/threaded.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+class Stacker:
+
+    NULL_UUID = uuid.UUID(int=0)
+
+    def __init__(
+        self,
+        config: Config,
+        uid: uuid.UUID = NULL_UUID,
+        stack_name_prefix: str = "tCat",
```
Comment (andrew-glenn) [inline taskcat/_cfn/threaded.py]: `tCat` -> `tCaT`

File: taskcat/_cfn/threaded.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                for k, v in test.parameters.items()
+            ],
+            "tags": tags,
+            "test_name": test.name,
+        }
+        stacks = fan_out(Stack.create, partial_kwargs, test.regions, threads)
+        self.stacks += stacks
+
+    @staticmethod
+    def _cfn_format_parameters(parameters):
```
Comment (andrew-glenn) [inline taskcat/_cfn/threaded.py]: Is this used anywhere? I see an opportunity at L72.

File: taskcat/_common_utils.py [hunk 3]
```diff
@@ -27,20 +31,19 @@ def name_from_stack_id(stack_id):
...
-        domain = get_s3_domain(location, client_factory)
+        domain = get_s3_domain(location)
         url = f"https://{bucket}.s3-{location}.{domain}/{key}"
     return url
 
 
-def get_s3_domain(region, client_factory):
-    ssm_client = client_factory.get("ssm")
+def get_s3_domain(region, ssm_client=None):
+    ssm_client = ssm_client if ssm_client else boto3.client("ssm")
```
Comment (andrew-glenn) [inline taskcat/_common_utils.py]: Standalone `boto3.client` versus the `region.client("ssm")`  👎

File: taskcat/_s3_stage.py [hunk 4]
```diff
@@ -34,26 +34,10 @@ def stage_in_s3(config):
...
 
-    # Create the bucket objects first!
     for test in config.tests.values():
         for region in test.regions:
             bucket_set.add(region.s3bucket)
-    # Sync!
     for bucket in bucket_set:
-        try:
-            bucket.create()
-        except Exception as e:
```
Comment (andrew-glenn) [inline taskcat/_s3_stage.py]: Why did you remove the exception handling?

**Comment by andrew-glenn** [comment]
Also see this: 
- https://github.com/aws-quickstart/taskcat/issues/318

File: taskcat/_s3_stage.py (hunk 4 above)
Comment (jaymccon) [reply taskcat/_s3_stage.py]: shouldn't be using `TaskcatException` for catch-all exceptions

File: taskcat/_common_utils.py (hunk 3 above)
Comment (jaymccon) [reply taskcat/_common_utils.py]: as this is querying public parameters, it does not require any specific account/region

Comment (andrew-glenn) [reply taskcat/_common_utils.py]: Yes, but the default credential set could simply have enough permissions to assume a role.

File: taskcat/_config.py [hunk 5]
```diff
@@ -208,18 +207,18 @@ def _assign_account_buckets(self):
                     bucket_dict,
                     config=self,
                     name=self.s3bucket.name,
-                    client=test_region.client,
+                    client=test_region.client("s3"),
```
Comment (avattathil) [inline taskcat/_config.py]: This call return a session correct ? `test_region.client("s3")  `
So this would make it a bit clearer
session = `test_region.client("s3")`

File: taskcat/_common_utils.py (hunk 3 above)
Comment (jaymccon) [reply taskcat/_common_utils.py]: agree that the default set should be fine, but then I need to pass in a client factory instance, which (unless I'm missing something) I may not have at this stage of execution.

File: taskcat/_config.py (hunk 5 above)
Comment (jaymccon) [reply taskcat/_config.py]: `s3bucket.client` (should) contain a boto3 client (equivalent of `boto3.client('s3')`), not session, wanted to make it an actual s3 client so that the consumer is not left with the task of ensuring the correct region.

**Comment by jaymccon** [comment]
> Also see this:
> 
> * [aws-quickstart/taskcat#318](https://github.com/aws-quickstart/taskcat/issues/318)

not importing requests from botocore anywhere in this pr, only remaining imports are from legacy parts of the code-base due for replacement

File: taskcat/_config.py (hunk 5 above)
Comment (andrew-glenn) [reply taskcat/_config.py]: @avattathil - Sorry, Jay is right. My bad for leading you in the wrong direction.

File: taskcat/_cfn/stack.py [hunk 6]
```diff
@@ -3,44 +3,57 @@
...
-    CREATE_COMPLETE = ["CREATE_COMPLETE"]
-    CREATE_FAILED = [
+    COMPLETE = ["CREATE_COMPLETE", "UPDATE_COMPLETE", "DELETE_COMPLETE"]
+    IN_PROGRESS = [
+        "CREATE_IN_PROGRESS",
+        "DELETE_IN_PROGRESS",
+        "UPDATE_IN_PROGRESS",
+        "UPDATE_COMPLETE_CLEANUP_IN_PROGRESS",
+    ]
+    FAILED = [
```
Comment (andrew-glenn) [inline taskcat/_cfn/stack.py]: Why was this reduced to COMPLETE/IN_PROGRESS/FAILED? 

I can see many scenarios where knowing a DELETE_FAILED occured (verus CREATE_FAILED) is useful.

File: taskcat/_cfn/stack.py [hunk 7]
```diff
@@ -50,59 +63,73 @@ class Capabilities:
...
+        self.key: str = param_dict["ParameterKey"]
+        self.value: str = ""
+        self.raw_value: str = ""
+        self.use_previous_value: bool = False
+        self.resolved_value: str = ""
+        if "ParameterValue" in param_dict.keys():
+            self.value = param_dict["ParameterValue"]
+        if "UsePreviousValue" in param_dict.keys():
+            self.use_previous_value = param_dict["UsePreviousValue"]
+        if "ResolvedValue" in param_dict.keys():
```
Comment (andrew-glenn) [inline taskcat/_cfn/stack.py]: `ResolvedValue` ?
`UsePreviousValue` ?

I don't recall these being available in this form in the v0.8 branch. What's the mechanism for addressing this with in-config params? (inline)

File: taskcat/_cfn/stack.py [hunk 8]
```diff
@@ -169,36 +228,51 @@ def __init__(
...
     ) -> "Stack":
-        cfn_client = template.client_factory_instance.get(
-            "cloudformation", region=region
-        )
+        uuid = uuid if uuid else uuid4()
+        cfn_client = region.client("cloudformation")
         parameters = [p.dump() for p in parameters] if parameters else []
         tags = [t.dump() for t in tags] if tags else []
+        if isinstance(region.s3bucket, S3Bucket):
+            bucket_name: str = region.s3bucket.name
```
Comment (andrew-glenn) [inline taskcat/_cfn/stack.py]: :+1:

File: taskcat/_cfn/threaded.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        tags += [
+            Tag(t)
+            for t in self.tags
+            if t.key not in ["taskcat-project-name", "taskcat-test-name", "taskcat-id"]
+        ]
+        fan_out(self._create_stacks_for_test, {"tags": tags}, tests, threads)
+
+    def _create_stacks_for_test(self, test, tags, threads: int = 32):
+        stack_name = "{}-{}-{}-{}".format(
+            self.stack_name_prefix, self.project_name, test.name, self.uid.hex
```
Comment (andrew-glenn) [inline taskcat/_cfn/threaded.py]: Suggestion:
```
name_section_list = []
for name_section in [self.stack_name_prefix, self.project_name, test.name, self.uid.hex]:
  if name_section:
     name_section_list.append(name_section)
stack_name = name_section_list.join('-')
```

This would automagically handle `self.stack_name_prefix == None`

File: taskcat/_config_types.py [hunk 10]
```diff
@@ -143,5 +147,161 @@ def client(self, service):
         )
         return service_session
 
+    def set_partition(self):
+        if not self._credset_name:
```
Comment (andrew-glenn) [inline taskcat/_config_types.py]: `self._credset_name` -> `self.credset_name`

File: taskcat/_cfn/stack.py (hunk 7 above)
Comment (jaymccon) [reply taskcat/_cfn/stack.py]: these are part of the cfn describe-stack api response, and will only be populated for stacks imported from existing, can consider them read only (ie. you cannot specify them in a config).

They have no practical use today, just including them so as to map the api as closesly as possible, maybe one day we find a use for them.

Comment (jaymccon) [reply taskcat/_cfn/stack.py]: UsePreviousValue is not exactly read-only, it can be specified when doing a stack update

File: taskcat/_cfn/stack.py (hunk 6 above)
Comment (jaymccon) [reply taskcat/_cfn/stack.py]: you will know specific status (and can query it):

```
s.stacks[0].status
Out[55]: 'CREATE_FAILED'
```

These are convenience groupings for the cases when you only want to know the general state. For example, line 390, where there is no point in continuing to poll `describe_stack` for a stack that's not in an `IN_PROGRESS` state
